### DISCUSSION TITLE
Idiomatic way to send exceptions in Shadow Endpoints to Sentry

### DISCUSSION THREAD

**Discussion by georgecrawford** [sveltejs/kit#3768] [Q&A] [answered]
Hi,

I'm trying to add Sentry error handling to my SvelteKit app. Since implementing shadow endpoints (which are awesome!) I can't work out where to call `Sentry.captureException()` for any errors which are thrown during the `get()` function of an endpoint. If I try to do it just after the `const response = await resolve(event);` in the `handle()` hook, the response is already a rendered version of my `__error.svelte` page. 

For now, I've added this hacky code to `__error.svelte`:

```js
<script context="module">
    import {browser} from '$app/env';

    export async function load({
        error,
        status,
        url,
        params,
        props,
    }) {
        if (!browser && !error.sentry) {
            const {captureException} = await import('$lib/errors/server.js');

            // NB: this calls through to Sentry.captureException
            const sentry = captureException({
                error,
                extra: {url, params, props},
            });
            Object.assign(error, {sentry});
        }

        return {
            props: {
                error,
                status,
            },
        };
    }
</script>
```

Problems:

- I have to ensure I only run this on the server
- `__error.svelte` can be defined anywhere, so I can't be sure that I'll _always_ correctly capture an exception and send to Sentry
- a Svelte file seems like completely the wrong place to do this. But there doesn't seem to be a place in `hooks.js` where I can see the error is thrown before SvelteKit starts to render error pages for me.

It seems like I need a new type of hook. `handleError` would seem logical, and the docs suggest so:
> handleError is only called in the case of an uncaught exception. It is not called when pages and endpoints explicitly respond with 4xx and 5xx status codes.

...but this isn't called when I throw an Error in an endpoint's `get()` function. Should it be?

**Comment by mustofa-id** [comment]
> this isn't called when I throw an Error in an endpoint's `get()` function

same here when using `shadow endpoint`, but It called when using `page load`.

**Reply by georgecrawford** [reply]
https://github.com/sveltejs/kit/issues/3807 is now fixed and released!

**Reply by mustofa-id** [reply]
Thank you for the fix!

**Comment by georgecrawford** [comment, marked as answer]
I've raised an issue: https://github.com/sveltejs/kit/issues/3807
